### PR TITLE
Fix wait() logic wrt timeout

### PR DESCRIPTION
**Pull request by AdamIsrael**
This PR addresses lp:1430488. 

The `waiter.wait` blocks for the full duration of the timeout, and `helpers.timeout` creates a signal-based timeout that's triggered when the timeout expires. This lead to `juju_agent()` being run against a unit still installing, which would throw an exception.

I made `waiter.wait` block for a smaller, arbitrary value of time (15 seconds), check the unit status, and repeat until it's ready or the timeout is expired.

### CODE REVIEW

File: amulet/sentry.py [hunk 1]
```diff
@@ -254,15 +255,16 @@ def wait_for_status(self, juju_env, services, timeout=300):
             raise
 
     def wait(self, timeout=300):
-        #for unit in self.unit:
         ready = False
         try:
+            now = time.time()
             with helpers.timeout(timeout):
                 # Make sure we're in a 'started' state across the board
```
Comment (mbruzek) [inline amulet/sentry.py]: The previous code does not actually check for "started" here.  I would recommend removing this comment to reflect the code.

**Comment by mbruzek** [comment]
@AdamIsrael I am not a reviewer for amulet code base, just curious really.  I looked this over and it appears the side effect is the code will check status more frequently, but does seem to be better waiting logic.

**Comment by AdamIsrael** [comment]
Thanks for the feedback Matt. I'm no longer convinced that this new logic fixes the underlying issue. I'll update the pull request after I've done further testing.

**Comment by AdamIsrael** [comment]
Hey @mbruzek, thanks for the feedback! I've dug into the issue deeper and updated this pull request.

There were two issues:

wait_for_status only checked to see if a unit had a public-address before considering it to be ready, but at least one provider (amazon) returns the public-address while the unit is still being allocated. This led to a condition where calling add_unit() would return before the unit was available, and subsequent calls against the unit could fail.

File: amulet/sentry.py [hunk 2]
```diff
@@ -235,7 +240,7 @@ def wait_for_status(self, juju_env, services, timeout=300):
                     ready = True
                     status = waiter.status(juju_env)
                     for service in services:
-                        if not 'units' in status['services'][service]:
+                        if 'units' not in status['services'][service]:
```
Comment (chuckbutler) [inline amulet/sentry.py]: Yoda conditionals :D

File: amulet/sentry.py [hunk 3]
```diff
@@ -254,24 +261,26 @@ def wait_for_status(self, juju_env, services, timeout=300):
...
-        #for unit in self.unit:
         ready = False
         try:
             with helpers.timeout(timeout):
                 # Make sure we're in a 'started' state across the board
-                waiter.wait(timeout=timeout)
-                while not ready:
+                now = time.time()
+                while not ready and now + timeout < time.time():
+                    waiter.wait(timeout=15)
```
Comment (chuckbutler) [inline amulet/sentry.py]: This is arbitrary right? or is there science being 15 seconds

File: amulet/sentry.py [hunk 4]
```diff
@@ -254,24 +261,26 @@ def wait_for_status(self, juju_env, services, timeout=300):
...
-                        # Check if we have a hook key and it's not None
-                        if status is None:
-                            ready = False
-                            break
-                        if 'hook' in status and status['hook']:
-                            ready = False
-                            break
-                        else:
-                            ready = True
+                        if unit['agent-state'] == 'started':
```
Comment (chuckbutler) [inline amulet/sentry.py]: :+1: on finding the status as started

File: amulet/sentry.py [hunk 5]
```diff
@@ -254,24 +261,26 @@ def wait_for_status(self, juju_env, services, timeout=300):
...
+                            status = self.unit[unit].juju_agent()
+
+                            # Check if we have a hook key and it's not None
+                            if status is None:
+                                ready = False
+                                break
+                            if 'hook' in status and status['hook']:
```
Comment (chuckbutler) [inline amulet/sentry.py]: This is wrt the new extended status support thats coming right? +1 again on this. very nice

**Comment by chuckbutler** [comment]
I have a few comments for clarity, but over this LGTM

File: amulet/sentry.py (hunk 3 above)
Comment (AdamIsrael) [inline amulet/sentry.py]: Totally arbitrary. I thought about calculating a sane value based on timeout, but 15 seemed sane enough.

File: amulet/sentry.py (hunk 5 above)
Comment (AdamIsrael) [inline amulet/sentry.py]: If you mean the 'hook' in status, that was existing. I'm not exactly sure what the use case for it is, so I left it as is.

File: amulet/sentry.py [hunk 6]
```diff
@@ -254,24 +261,26 @@ def wait_for_status(self, juju_env, services, timeout=300):
...
     def wait(self, timeout=300):
-        #for unit in self.unit:
         ready = False
         try:
             with helpers.timeout(timeout):
                 # Make sure we're in a 'started' state across the board
-                waiter.wait(timeout=timeout)
-                while not ready:
+                now = time.time()
+                while not ready and now + timeout < time.time():
```
Comment (tvansteenburgh) [inline amulet/sentry.py]: I'm not sure you want the extra time checking here. Clients are expecting this method (wait()) to raise if units aren't ready within the timeout. I think just `while not ready` is better here.

**Comment by AdamIsrael** [comment]
While reviewing @tvansteenburgh's comments, I discovered that I wasn't solving the underlying problem.

`UnitSentry._run` relied on the default `juju run` timeout of 5 minutes, and this method blocks if other hook contexts are executing. In the case of the ubuntu-repository-cache charm, that hook context was doing a full rsync of the Ubuntu archive and takes longer than 5 minutes to run.

`Talisman.wait_for_status` now waits until all units are in a 'started' state and have a public-address assigned, so the `Talisman.wait` method will now honor the timeout parameter when waiting for the `juju_agent` method to be run against the unit.

PR #68 adds a test for calling `add_unit` against an already deployed service.

**Pull request merged**